### PR TITLE
CI: add job with external tests (temporarily krb5 and gost_engine only)

### PR DESCRIPTION
**Pull request by t8m**
Update gost-engine submodule.
Update pyca-cryptography submodule.

Fix condition for skipping krb5 test.

As the external tests for boringssl and pyca are failing currently, this is WIP.

### CODE REVIEW

**Comment by t8m** [comment]
One option would be to temporarily run just the gost and krb5 tests for now.

**Comment by beldmit** [comment]
1. To run the Perl gost-engine tests, it's necessary to install Test::V0. When it is available during the configuration phase, the perl tests will run automagically.
2. Could you also add the TCL tests to the gost-engine tests? They were not integrated into my original test scenario.

**Comment by t8m** [comment]
>     1. To run the Perl gost-engine tests, it's necessary to install Test::V0. When it is available during the configuration phase, the perl tests will run automagically.
> 
>     2. Could you also add the TCL tests to the gost-engine tests? They were not integrated into my original test scenario.

Added. Did I do it correctly?

**Comment by t8m** [comment]
Hmm, apparently not. @beldmit any hints?

**Comment by beldmit** [comment]
You use the default config test that doesn't load the gost engine at all

**Comment by beldmit** [comment]
Could you also try to run the tcl tests using the command from #14412?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -166,3 +166,24 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        submodules: recursive
+    - name: package installs
+      run: |
+        sudo apt-get update
+        sudo apt-get -yq install bison dejagnu gettext keyutils ldap-utils libldap2-dev libkeyutils-dev python3 python3-paste python3-pyrad slapd tcl-dev tcl-thread tcsh python3-virtualenv virtualenv
+    - name: install cpanm and Test2::V0
```
Comment (levitte) [inline .github/workflows/ci.yml:180]: I suggest adding "(for gost-engine testing)"

**Comment by t8m** [comment]
@beldmit, for some reason the make tcl_tests still fails for me and I do not see anything special in your PR that would not be already in the gost_engine.sh script as well.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (t8m) [reply .github/workflows/ci.yml]: Done

File: test/recipes/95-test_external_gost_engine_data/gost_engine.sh [hunk 2]
```diff
@@ -12,14 +12,21 @@
...
-O_LIB=`pwd`/$BLDTOP
+PWD="$(pwd)"
 
-export PATH=$O_EXE:$PATH
-export LD_LIBRARY_PATH=$O_LIB:$LD_LIBRARY_PATH
-export OPENSSL_ROOT_DIR=$O_LIB
+O_EXE="$PWD/$BLDTOP/apps"
+O_BINC="$PWD/$BLDTOP/include"
+O_SINC="$PWD/$SRCTOP/include"
+O_LIB="$PWD/$BLDTOP"
```
Comment (levitte) [inline test/recipes/95-test_external_gost_engine_data/gost_engine.sh:20]: Ouch.  If any of those are absolute paths, the failure will be glorious.  I'd suggest this instead:

``` bash
SRCTOP=$(cd $SRCTOP; pwd)
BLDTOP=$(cd $BLDTOP; pwd)
O_EXE=$BLDTOP/apps
O_BINC=$BLDTOP/include
O_SINC=$SRCTOP/include
O_LIB=$BLDTOP
```

**Comment by beldmit** [comment]
Let's temporary switch the TCL tests off. I remember that here be some dragons, so I'd better add them myself later.

File: test/recipes/95-test_external_gost_engine_data/gost_engine.sh [hunk 3]
```diff
@@ -37,9 +44,11 @@ cd $SRCTOP/gost-engine
 rm -rf build
 mkdir -p build
 cd build
-cmake ..
+cmake .. -DOPENSSL_ROOT_DIR="$OPENSSL_ROOT_DIR"
```
Comment (levitte) [inline test/recipes/95-test_external_gost_engine_data/gost_engine.sh:49]: Argh!  No!  Don't build in the source tree!

As a matter of fact, there's no need to move anywhere, as this script should be standing in `test-runs/test_external_gost_engine` already.  So, the bunch of lines from `cd $SRCTOP/gost-engine` to here can be replaced with this:

``` bash
cmake $SRCTOP/gost-engine -DOPENSSL_ROOT_DIR="$OPENSSL_ROOT_DIR"
```

**Review by levitte — CHANGES_REQUESTED** [review]

File: test/recipes/95-test_external_gost_engine_data/gost_engine.sh (hunk 3 above)
Comment (t8m) [reply test/recipes/95-test_external_gost_engine_data/gost_engine.sh:49]: Yeah, I just was not changing it from what was there before already. But I can try this.

**Comment by t8m** [comment]
OK the gost_engine test now passes including the tcl test. I'll try to add the improvements suggested by Richard now.

File: test/recipes/95-test_external_gost_engine_data/gost_engine.sh (hunk 3 above)
Comment (beldmit) [reply test/recipes/95-test_external_gost_engine_data/gost_engine.sh:49]: > Argh! No! Don't build in the source tree!
> 
> As a matter of fact, there's no need to move anywhere, as this script should be standing in `test-runs/test_external_gost_engine` already. So, the bunch of lines from `cd $SRCTOP/gost-engine` to here can be replaced with this:
> 
> ```shell
> cmake $SRCTOP/gost-engine -DOPENSSL_ROOT_DIR="$OPENSSL_ROOT_DIR"
> ```

I'm sorry. I partially look at all this stuff as at Ancient Magic, and prefer not to dive in it...

**Comment by t8m** [comment]
Pulling out of WIP.

@levitte comments addressed. The krb5 and gost_engine tests are running correctly. The other two have separate issues opened.

**Comment by beldmit** [comment]
Nice work!
Isn't it worth making each external test a separate phase?

**Review by beldmit — APPROVED** [review]
LGTM. The only question is raised.

**Comment by t8m** [comment]
That makes some sense as you wouldn't have to look into the log to see which external test failed.

**Comment by levitte** [comment]
So wait, the tcl tests work? Cool!
(they were failing incessantly over att gost-engine for a while)

**Comment by beldmit** [comment]
> So wait, the tcl tests work? Cool!
> (they were failing incessantly over att gost-engine for a while)

The last (for now) TCL test failure was eliminated by @DDvO some time ago, and I don't want to introduce the new ones.

**Comment by t8m** [comment]
@beldmit I've separated the tests, please re-approve

**Review by beldmit — APPROVED** [review]
LGTM

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -185,5 +185,11 @@ jobs:
       run: ./config --strict-warnings --debug no-afalgeng enable-rc5 enable-md2 enable-ssl3 enable-ssl3-method enable-weak-ssl-ciphers enable-zlib enable-ec_nistp_64_gcc_128 enable-external-tests && perl configdata.pm --dump
     - name: make
       run: make -s -j4
-    - name: make test
-      run: BORING_RUNNER_DIR=$(pwd)/boringssl/ssl/test/runner make test TESTS="test_external_gost_engine test_external_krb5"
+    - name: make test_external_gost_engine
```
Comment (levitte) [inline .github/workflows/ci.yml:188]: `name: test external gost-engine`

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -185,5 +185,11 @@ jobs:
...
+      run: make test TESTS="test_external_gost_engine"
+    - name: make test_external_krb5
```
Comment (levitte) [inline .github/workflows/ci.yml:190]: `name: test external krb5`

**Review by levitte — CHANGES_REQUESTED** [review]
... and so on

**Comment by t8m** [comment]
> ... and so on

OK, you nitpicker. :grin: 

Fixed. @beldmit still OK?

**Review by levitte — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews!

**Pull request closed without merging**